jni/src/test/java/com/horizen/certnative/NaiveThresholdSignatureWKeyRotationProofTest.java

```diff
@@ -149,9 +119,199 @@ private void testCreateVerifyRandomProof(int numCustomFields, String snarkPkPath
+        testKeysSignaturesList.getUpdatedSigningKeysMkSignatures()[0] = masterKeyPairList.get(0).signMessage(newSigningKeyHash);
+        testKeysSignaturesList.getUpdatedSigningKeys()[0] = newSigningKeyPair.getPublicKey();
+
+        aWithdrawalCertificate.getCustomFields()[0] = testKeysSignaturesList.getUpdatedKeysRootHash(keyCount);
+
+        List<SchnorrSignature> signatureList = createSignatures(aWithdrawalCertificate, signingKeyPairList);
+
+        Optional<String> failingConstraint = NaiveThresholdSignatureWKeyRotation.debugCircuit(
+            testKeysSignaturesList,
+            aWithdrawalCertificate,
+            Optional.of(aPrevWithdrawalCertificate),
+            signatureList,
+            keyCount,
+            threshold,
+            testGenesisKeyRootHash
+        );
+        if (failingConstraint.isPresent()) {
+            throw new Exception(failingConstraint.get());
+        }
+        assertFalse(failingConstraint.isPresent());
+    }
+
+    @Test
+    public void testMasterKeyRotation() throws Exception {
+        Random r = new Random();
+
+        // Setup
+        scId = FieldElement.createRandom();
+        prevScId = FieldElement.createRandom();
+        List<SchnorrKeyPair> signingKeyPairList = setupKeyPairList(keyCount);
+        List<SchnorrKeyPair> masterKeyPairList = setupKeyPairList(keyCount);
+        ValidatorKeysUpdatesList testKeysSignaturesList = setupValidatorKeysUpdatesList(
+            signingKeyPairList, masterKeyPairList
+        );
+
+        FieldElement testPrevValidatorsKeysRoot = testKeysSignaturesList.getKeysRootHash(keyCount);
+        FieldElement testValidatorsKeysRoot = testKeysSignaturesList.getUpdatedKeysRootHash(keyCount);
+        FieldElement testGenesisKeyRootHash = testPrevValidatorsKeysRoot;
+        FieldElement constant = NaiveThresholdSignatureWKeyRotation.getConstant(testGenesisKeyRootHash, threshold);
+
+        // Create dummy Backward Transfers
+        List<BackwardTransfer> btList = getBackwardTransferList(backwardTransferCount);
+        List<BackwardTransfer> prevBtList = getBackwardTransferList(backwardTransferCount);
+        List<FieldElement> customFields = new ArrayList<>();
+        List<FieldElement> prevCustomFields = new ArrayList<>();
+        // Generate random custom fields if requested
+        customFields.add(testValidatorsKeysRoot);
+        prevCustomFields.add(testPrevValidatorsKeysRoot);
+
+        WithdrawalCertificate aWithdrawalCertificate = new WithdrawalCertificate(
+            scId, epochNumber, btList, (long)keyCount, testValidatorsKeysRoot,
+            ftMinAmount, btrFee, customFields
+        );
+        WithdrawalCertificate aPrevWithdrawalCertificate = new WithdrawalCertificate(
+            prevScId, prevEpochNumber, prevBtList, (long)keyCount, testPrevValidatorsKeysRoot,
+            prevFtMinAmount, prevBtrFee, prevCustomFields
+        );
+
+        // Rotate a signing key
```

Comment:
Minor: update comment to `Rotate a master key`

Reply:
Done